### ISSUE TITLE
Unable to access Vault secrets due to incorrect AWS region

### ISSUE THREAD

**Issue by dipack95** [hashicorp/vault-lambda-extension#18]
I'm trying to use this Lambda extension to fetch secrets from Vault, before the code in Lambda function runs. This is the error I'm receiving:

```text
START RequestId: 21b2eabf-0f3e-4ff2-8dca-932a6041e245 Version: $LATEST
8cd0a2c4-942c-4cbc-b6a2-0de422a07ba6[vault-lambda-extension] 2020/11/25 23:05:44 Initialising
[vault-lambda-extension] 2020/11/25 23:05:44 attemping Vault login...
[vault-lambda-extension] 2020/11/25 23:05:45 error getting client: Error making API request.

URL: PUT https://vault.xx.com/v1/auth/aws/login
Code: 400. Errors:

* error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>SignatureDoesNotMatch</Code>
    <Message>Credential should be scoped to a valid region, not 'us-east-1'. </Message>
  </Error>
  <RequestId>7df68af3-8edc-44a3-b75c-230b1b36dccc</RequestId>
</ErrorResponse>
EXTENSION	Name: vault-lambda-extension	State: Registered	Events: [INVOKE,SHUTDOWN]
END RequestId: 21b2eabf-0f3e-4ff2-8dca-932a6041e245
REPORT RequestId: 21b2eabf-0f3e-4ff2-8dca-932a6041e245	Duration: 5002.45 ms	Billed Duration: 5100 ms	Memory Size: 128 MB	Max Memory Used: 37 MB	
XRAY TraceId: 1-5fbee346-546a8a1d79b0e587206cea7f	SegmentId: 75925ef155507a5a	Sampled: true	
RequestId: 21b2eabf-0f3e-4ff2-8dca-932a6041e245 Error: exit status 1
Extension.Crash
```

This is reproducible on my local machine as well:

```bash
~/projects  aws-vault exec identity -- vault login --method=aws role=developer_role                                                                                                                                 
Error authenticating: Error making API request.

URL: PUT https://vault.xx.com/v1/auth/aws/login
Code: 400. Errors:

* error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>SignatureDoesNotMatch</Code>
    <Message>Credential should be scoped to a valid region, not 'us-east-1'. </Message>
  </Error>
  <RequestId>x-a0fd-4e52-a9d6-x</RequestId>
</ErrorResponse>

~/projects  aws-vault exec identity -- vault login --method=aws role=developer_role region=eu-west-1                                                                                                                      
Error authenticating: Error making API request.

URL: PUT https://vault.xx.com/v1/auth/aws/login
Code: 400. Errors:

* error looking up full ARN of entity &{aws 5218125xx user  dipack95 }: error fetching user "dipack95": AccessDenied: User: arn:aws:sts::5903x:assumed-role/vault-clusterx/i-x is not authorized to perform: iam:GetUser on resource: user dipack95
        status code: 403, request id: x-a8ef-4699-9e92-x
```

The second command, running with the `region=eu-west-1` flag works fine, as it's throwing a regular AWS error for access denied, which is expected, but as you can see, the first command defaults to `us-east-1`, throwing the same error as the Lambda function.

Is there a solution to this?

**Comment by tomhjp** [comment]
Hi @dipack95, thanks for raising this. It sounds to me like there is a mismatch in regions between where the STS request is generated and which region's STS endpoint Vault uses to verify. I think we probably need to add something like `d["region"] = os.Getenv("AWS_REGION")` to the data for the API login call [here](https://github.com/hashicorp/vault-lambda-extension/blob/main/vault/client.go#L47).

However, I would like to be able to reproduce your issue first to confirm the fix. Please could you help me with fuller steps to reproduce? I tried modifying the quick-start guide to place the Lambda in `eu-west-2` but everything else in `eu-west-1` and it worked fine for me, which suggests I'm still missing something.

**Comment by dipack95** [comment]
@tomhjp I've been using Serverless to create my Lambda functions, and attach the lambda extension for Vault, so let me see if I can't have a reproducible example config for you!

**Comment by dipack95** [comment]
@tomhjp Here's a reproduction of the issue in the parent post:

### `serverless.yml`
```yaml
service: hashicorp-vault-extension-region-issue

provider:
  name: aws
  runtime: nodejs12.x
  region: eu-west-1
  timeout: 10
  memorySize: 128
  tracing:
    lambda: true

functions:
  helloWorldHandler:
    name: hello-world-handler
    handler: handler.handler
    layers:
      - { Ref: CertsLambdaLayer }
      - arn:aws:lambda:eu-west-1:634166935893:layer:vault-lambda-extension:6
    environment:
      VAULT_ADDR: https://vault.dev.trancsend.com
      VAULT_AUTH_PROVIDER: aws
      VAULT_AUTH_ROLE: backend_role
      VAULT_SECRET_PATH: kv/integrations/args
      VAULT_SECRET_FILE: /opt/vault/secrets.json
      VAULT_CACERT: /opt/ca-crt.pem

layers:
  certs:
    path: ./ssl
```

### `handler.js`
```javascript
const { readFileSync, readdirSync, existsSync } = require('fs');

function readVaultSecretsJSON() {
  const secretsFile = '/opt/vault/secrets.json';
  console.log({ files: readdirSync('/opt') });
  if (!existsSync(secretsFile)) {
    throw new Error(`No file ${secretsFile} found!`);
  }
  const contents = readFileSync(secretsFile, { encoding: 'utf-8' });
  console.log(contents);
  return JSON.parse(contents);
}

module.exports.handler = (event) => {
    console.log('Hello world!');
    console.log('Secrets', readVaultSecretsJSON());
};
```

### Error Message on AWS Console
```text
START RequestId: x Version: $LATEST
x[vault-lambda-extension] 2020/12/01 17:40:58 Initialising
[vault-lambda-extension] 2020/12/01 17:40:58 attemping Vault login...
[vault-lambda-extension] 2020/12/01 17:40:59 error getting client: Error making API request.

URL: PUT https://vault.xx.com/v1/auth/aws/login
Code: 400. Errors:

* error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>SignatureDoesNotMatch</Code>
    <Message>Credential should be scoped to a valid region, not 'us-east-1'. </Message>
  </Error>
  <RequestId>x</RequestId>
</ErrorResponse>
EXTENSION	Name: vault-lambda-extension	State: Registered	Events: [INVOKE,SHUTDOWN]
END RequestId: x
REPORT RequestId: xDuration: 4321.01 ms	Billed Duration: 4322 ms	Memory Size: 128 MB	Max Memory Used: 35 MB	
XRAY TraceId: x SegmentId: x	Sampled: true	
RequestId: x Error: exit status 1
Extension.Crash
```

**Comment by tomhjp** [comment]
Hi @dipack95, sorry for the long delay here, but thanks for providing all of that config. Unfortunately I'm still not able to repro. For simplicity and getting to a minimum repro, I stripped out the TLS from your yaml, and also replaced `/opt` with `/tmp` because `/opt` is readonly in the lambda execution environment, then I ran vault locally with ngrok, but otherwise everything is deployed and working as in your configs. I'm really at a loss to what's different about your setup that's causing the issue, but presumably something about your Vault config? Are you able to include some steps to setup Vault that make this reproduce?

### `serverless.yml`
```yaml
service: vle-18 

provider:
  name: aws
  runtime: nodejs12.x
  region: eu-west-1
  timeout: 10
  memorySize: 128
  tracing:
    lambda: true

functions:
  helloWorldHandler:
    name: hello-world-handler
    handler: handler.handler
    layers:
      - arn:aws:lambda:eu-west-1:634166935893:layer:vault-lambda-extension:6
    environment:
      VAULT_ADDR: http://xx.ngrok.io
      VAULT_AUTH_PROVIDER: aws
      VAULT_AUTH_ROLE: backend_role
      VAULT_SECRET_PATH: kv/integrations/args
      VAULT_SECRET_FILE: /tmp/vault/secrets.json
```

### `handler.js`
```javascript
const { readFileSync, readdirSync, existsSync } = require('fs');

function readVaultSecretsJSON() {
  const secretsFile = '/tmp/vault/secrets.json';
  console.log({ files: readdirSync('/tmp') });
  if (!existsSync(secretsFile)) {
    throw new Error(`No file ${secretsFile} found!`);
  }
  const contents = readFileSync(secretsFile, { encoding: 'utf-8' });
  console.log(contents);
  return JSON.parse(contents);
}

module.exports.handler = (event) => {
    console.log('Hello world!');
    console.log('Secrets', readVaultSecretsJSON());
};
```

### Vault setup

```bash
export VAULT_ADDR=http://127.0.0.1:8200;
export VAULT_TOKEN=root
VAULT_DEV_ROOT_TOKEN_ID=root vault server -dev &
ngrok http 8200 -log=stdout --bind-tls false &
vault secrets enable -version=1 kv
vault kv put integrations/args foo=bar
vault policy write backend_policy - <<EOF
path "kv/*" {
    capabilities = ["read", "list"]
}
EOF
vault auth enable aws
vault write -force auth/aws/config/client
vault write auth/aws/role/backend_role \
    auth_type=iam \
    policies=backend_policy \
    ttl=1 \
    resolve_aws_unique_ids=false \
    bound_iam_principal_arn="<ARN-of-the-function's-role>"
```

### Logs

```
2021-01-04T13:48:05.777Z	e4189930-de10-4873-8766-eb97eef30579	INFO	Hello world!
2021-01-04T13:48:05.817Z	e4189930-de10-4873-8766-eb97eef30579	INFO	{ files: [ 'vault' ] }
2021-01-04T13:48:05.817Z	e4189930-de10-4873-8766-eb97eef30579	INFO	{
    "request_id": "f1c9a09c-4750-7ad5-a0db-f21017720c7d",
    "lease_id": "",
    "lease_duration": 2764800,
    "renewable": false,
    "data": {
        "foo": "bar"
    },
    "warnings": null
}
2021-01-04T13:48:05.817Z	e4189930-de10-4873-8766-eb97eef30579	INFO	Secrets {  request_id: 'f1c9a09c-4750-7ad5-a0db-f21017720c7d',  lease_id: '',  lease_duration: 2764800,  renewable: false,  data: { foo: 'bar' },  warnings: null}
```

**Comment by dmattia** [comment]
I have a similar problem, and can reproduce the error in this issue from your example by changing the line:

```bash
vault write -force auth/aws/config/client
```

to be:

```bash
vault write auth/aws/config/client \
    sts_endpoint="https://sts.eu-west-1.amazonaws.com" \
    sts_region="eu-west-1"
```

At that point, I see the same issue as @dipack95. If I then type `vault write -force auth/aws/config/client` and redeploy serverless to get a new lambda executor, the secrets print out successfully.

**Comment by dmattia** [comment]
Hello @tomhjp! Friendly ping on this issue. Could you please let us know if this is something being looked into, or if we should have a separate plan for the time being?

**Comment by tomhjp** [comment]
Sorry for the delay on this. Thanks for the repro steps @dmattia, I've now managed to reproduce this locally, and tested a fix, which basically involves creating the `sts` session with regional settings like so:

```golang
ses := session.Must(session.NewSession(&aws.Config{
	Region:              aws.String("eu-west-1"),
	STSRegionalEndpoint: endpoints.RegionalSTSEndpoint,
}))
```

I think the proper fix will be to add an environment variable to configure that the extension should use a specific region's `sts` endpoint, as users can have their Vault IAM auth mount configured for either the global or the regional endpoint. I hope to get a fix for this landed in time for the next release in the coming weeks.

**Comment by tomhjp** [comment]
Actually on closer inspection of the AWS SDK, you should be able to fix this without a patch to `vault-lambda-extension`. For logging into Vault, we currently do `ses := session.Must(session.NewSession())`, which loads all the AWS configuration from environment variables. `AWS_REGION` should already be set to the correct region by default for you ([docs](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime)), so if you set `AWS_STS_REGIONAL_ENDPOINTS=regional` ([docs](https://docs.aws.amazon.com/credref/latest/refdocs/setting-global-sts_regional_endpoints.html#setting-sts_regional_endpoints-details)) for your Lambda, it should correctly use the regional STS endpoints for generating a signed request that will match up to your Vault config.

I've just tested the above with my local repro, and can confirm it did work. Do let me know if it doesn't work for you though.

This isn't easy to discover, so I'll keep this issue open for now as a documentation improvement.

**Comment by dipack95** [comment]
Thanks for figuring it out, @tomhjp! I'm going to try out your suggestion sometime in the next couple of days, and see if it works for our case.

**Comment by tomhjp** [comment]
@dipack95 I'm pretty confident this issue is resolved with the documentation in #29, so I'm going to close for now, but feel free to comment again if you still have the same problem and I can reopen.
